File: bot/extensions/queue.py
import hikari
import lightbulb

from bot.library.checks import valid_user_voice, player_playing
from bot.library.classes.choice import AutocompleteChoice 
from bot.library.classes.sources import Spotify, Deezer
from bot.utils import player_bar, format_time

DELETE_AFTER = 60
plugin = lightbulb.Plugin('Queue', 'Queue commands')


@plugin.command()
@lightbulb.add_checks(
    lightbulb.guild_only, player_playing
)
@lightbulb.command('now', 'Display current track')
@lightbulb.implements(lightbulb.SlashCommand)
async def now(ctx: lightbulb.Context) -> None:
    """Display current track"""

    player = plugin.bot.d.lavalink.player_manager.get(ctx.guild_id)
    current = player.current

    desc = '[{}]({})\n{}\n{}\n\nRequested: <@!{}>\n'.format(
        current.title, current.uri, current.author,
        player_bar(player), current.requester)
    
    if player.queue:
        track = player.queue[0]
        desc += '\n**Up next:**\n[{}]({}) `{}`'.format(
            track.title, track.uri,
            'LIVE' if track.stream else format_time(track.duration))
        if track.source_name in (source.source_name.lower() for source in (Deezer, Spotify)):
            desc += f' {track.author}'

    await ctx.respond(
        embed=hikari.Embed(
            title = '🎵 Now Playing',
            description=desc).set_thumbnail(current.artwork_url))

@plugin.command()
@lightbulb.add_checks(
    lightbulb.guild_only, player_playing,
)
@lightbulb.command('queue', 'Display the next 10 tracks in queue')
@lightbulb.implements(lightbulb.SlashCommand)
async def queue(ctx : lightbulb.Context) -> None:
    """Display next (max 10) tracks in queue"""

    player = plugin.bot.d.lavalink.player_manager.get(ctx.guild_id)
    current = player.current

    desc = '[{}]({})\n{}\n{}\n\nRequested: <@!{}>\n'.format(
        current.title, current.uri, current.author,
        player_bar(player), current.requester)

    for i, track in enumerate(player.queue):
        if i == 0:
            desc += '\n**Up next:**'
        if i >= 10:
            break

        desc += '\n{}. [{}]({}) `{}`'.format(
            i + 1, track.title, track.uri,
            'LIVE' if track.stream else format_time(track.duration))
        if track.source_name in (source.source_name.lower() for source in (Deezer, Spotify)):
            desc += f' {track.author}'

    await ctx.respond(
        embed=hikari.Embed(
            title = '🎵 Queue',
            description = desc,
        ).set_thumbnail(current.artwork_url))

async def remove_autocomplete(option, interaction):
    
    player = plugin.bot.d.lavalink.player_manager.get(interaction.guild_id)
    if not player or not player.is_playing:
        return None
    return [AutocompleteChoice(f'{track.title[:60]} - {track.author[:20]}', str(i)) for i, track in enumerate(player.queue[:25])]

@plugin.command()
@lightbulb.add_checks(
    lightbulb.guild_only, valid_user_voice, player_playing
)
@lightbulb.option('track', 'Track to remove', required=True, autocomplete=remove_autocomplete)
@lightbulb.command('remove', 'Remove a track from queue')
@lightbulb.implements(lightbulb.SlashCommand)
async def remove(ctx: lightbulb.Context) -> None:
    """Remove a track from queue"""

    index = ctx.options.track
    if not index.isdigit():
        await ctx.respond('Queue empty - Nothing to remove!', flags=hikari.MessageFlag.EPHEMERAL)
        return

    player = plugin.bot.d.lavalink.player_manager.get(ctx.guild_id)
    popped_track = player.queue.pop(int(index))

    await ctx.respond(
        embed=hikari.Embed(
            description = f'Removed: [{popped_track.title}]({popped_track.uri})'),
            delete_after=DELETE_AFTER)


def load(bot: lightbulb.BotApp) -> None:
    bot.add_plugin(plugin)

def unload(bot: lightbulb.BotApp) -> None:
    bot.remove_plugin(plugin)

File: commands/queue.go
package commands

import (
	"MusicCatGo/musicbot"
	"fmt"

	"github.com/disgoorg/disgo/discord"
	"github.com/disgoorg/disgo/handler"
)

func (c *Commands) Now(data discord.SlashCommandInteractionData, e *handler.CommandEvent) error {

	player, ok := c.PlayerManager.GetPlayer(*e.GuildID())
	if !ok || !player.IsPlaying() {
		if sendErr := e.CreateMessage(discord.MessageCreate{
			Content: "Player is not playing.",
			Flags:   discord.MessageFlagEphemeral,
		}); sendErr != nil {
			musicbot.LogSendError(sendErr, e.GuildID().String(), e.User().ID.String(), true)
		}
		return nil
	}

	var (
		track    = player.Current()
		queue    = player.Queue()
		paused   = player.IsPaused()
		position = player.Position()
		userData UserData
	)
	_ = track.UserData.Unmarshal(&userData)

	content := fmt.Sprintf("[%s](%s)\n%s\n%s\n\nRequested: <@!%s>\n",
		track.Info.Title, *track.Info.URI, track.Info.Author,
		musicbot.PlayerBar(paused, track, position), userData.Requester)

	if len(queue) >= 1 {
		content += "\n**Up next:**"
		nextTrack := queue[0]
		var Playtime string
		if nextTrack.Info.IsStream {
			Playtime = "`LIVE`"
		} else {
			Playtime = musicbot.FormatTime(nextTrack.Info.Length)
		}
		content += fmt.Sprintf("\n[%s](%s) `%s`",
			nextTrack.Info.Title, *nextTrack.Info.URI, Playtime)

		if nextTrack.Info.SourceName == "deezer" || nextTrack.Info.SourceName == "spotify" {
			content += " " + nextTrack.Info.Author
		}
	}

	embed := discord.NewEmbedBuilder().
		SetTitle("Current").
		SetDescription(content).
		SetThumbnail(*track.Info.ArtworkURL)

	if sendErr := e.CreateMessage(discord.MessageCreate{
		Embeds: []discord.Embed{embed.Build()},
	}); sendErr != nil {
		musicbot.LogSendError(sendErr, e.GuildID().String(), e.User().ID.String(), false)
	}
	return nil
}

func (c *Commands) Queue(data discord.SlashCommandInteractionData, e *handler.CommandEvent) error {

	player, ok := c.PlayerManager.GetPlayer(*e.GuildID())
	if !ok || !player.IsPlaying() {
		if sendErr := e.CreateMessage(discord.MessageCreate{
			Content: "Player is not playing.",
			Flags:   discord.MessageFlagEphemeral,
		}); sendErr != nil {
			musicbot.LogSendError(sendErr, e.GuildID().String(), e.User().ID.String(), true)
		}
		return nil
	}

	var (
		track    = player.Current()
		queue    = player.Queue()
		paused   = player.IsPaused()
		position = player.Position()
		userData UserData
	)
	_ = track.UserData.Unmarshal(&userData)

	content := fmt.Sprintf("[%s](%s)\n%s\n%s\n\nRequested: <@!%s>\n",
		track.Info.Title, *track.Info.URI, track.Info.Author,
		musicbot.PlayerBar(paused, track, position), userData.Requester)

	limit := min(10, len(queue))
	for i, track := range queue[:limit] {
		if i == 0 {
			content += fmt.Sprintf("\n**Up next:** `%d track(s)`", len(queue))
		}

		var Playtime string
		if track.Info.IsStream {
			Playtime = "`LIVE`"
		} else {
			Playtime = musicbot.FormatTime(track.Info.Length)
		}
		content += fmt.Sprintf("\n%d. [%s](%s) `%s`",
			i+1, track.Info.Title, *track.Info.URI, Playtime)

		if track.Info.SourceName == "deezer" || track.Info.SourceName == "spotify" {
			content += " " + track.Info.Author
		}
	}

	embed := discord.NewEmbedBuilder().
		SetTitle("Queue").
		SetDescription(content).
		SetThumbnail(*track.Info.ArtworkURL)

	if sendErr := e.CreateMessage(discord.MessageCreate{
		Embeds: []discord.Embed{embed.Build()},
	}); sendErr != nil {
		musicbot.LogSendError(sendErr, e.GuildID().String(), e.User().ID.String(), false)
	}
	return nil
}
